File: pcc/py_runtime/src/py_timer_heap.c
/* py_timer_heap.c - scalable virtual-thread timer structure (C mirror).
 *
 * Mirrors pcc/vthread/timer_oracle.py::MinHeapTimerQueue exactly. See the
 * header for the algorithm and the semantics contract, and
 * docs/design/pcc-vthread-oracles.md for the design rationale (why a min-heap
 * with lazy cancellation, and not a timing wheel, for this first slice).
 *
 * Structure:
 *   * ``nodes`` is a 0-based binary min-heap of (deadline, seq, timer_id).
 *     Ordering key is (deadline, seq): sift-up / sift-down compare deadline
 *     first, breaking ties by the monotonic ``seq`` so equal-deadline entries
 *     pop in insertion (FIFO) order, matching the C sorted-list ``<= deadline``
 *     stable walk.
 *   * ``live`` is an open-addressing hash table timer_id -> (deadline, seq). It
 *     is the source of truth for membership and cancellation. A heap node is
 *     stale iff live has no entry for its id (cancelled/expired) or the entry's
 *     (deadline, seq) differs (the id was rescheduled to a newer slot).
 */
#include "py_timer_heap.h"

#include <stdlib.h>
#include <string.h>

#define PCC_TIMER_HEAP_MIN_CAP 8
#define PCC_TIMER_LIVE_MIN_CAP 8

/* ---- live-map (open addressing, linear probe) --------------------------- */

static uint64_t pcc_timer_hash_id(int64_t timer_id) {
    /* splitmix64 finalizer: good spread for sequential ids. */
    uint64_t x = (uint64_t)timer_id + 0x9e3779b97f4a7c15ULL;
    x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9ULL;
    x = (x ^ (x >> 27)) * 0x94d049bb133111ebULL;
    x = x ^ (x >> 31);
    return x;
}

/* Find the slot for timer_id. If found (state==1), returns its index and sets
 * *found=1. Otherwise returns the first insertion-eligible index (empty or
 * tombstone) and sets *found=0. Assumes live_cap > live_used (spare slot). */
static int64_t pcc_timer_live_probe(
    const PccTimerHeap *h,
    int64_t timer_id,
    int *found
) {
    int64_t mask = h->live_cap - 1;
    int64_t idx = (int64_t)(pcc_timer_hash_id(timer_id) & (uint64_t)mask);
    int64_t first_free = -1;
    for (;;) {
        PccTimerLiveSlot *slot = &h->live[idx];
        if (slot->state == 0) {
            *found = 0;
            return first_free >= 0 ? first_free : idx;
        }
        if (slot->state == 2) {
            if (first_free < 0) first_free = idx;
        } else if (slot->timer_id == timer_id) {
            *found = 1;
            return idx;
        }
        idx = (idx + 1) & mask;
    }
}

static int pcc_timer_live_rehash(PccTimerHeap *h, int64_t new_cap) {
    PccTimerLiveSlot *old = h->live;
    int64_t old_cap = h->live_cap;
    PccTimerLiveSlot *fresh = (PccTimerLiveSlot *)calloc(
        (size_t)new_cap, sizeof(PccTimerLiveSlot)
    );
    if (fresh == NULL) return -1;
    h->live = fresh;
    h->live_cap = new_cap;
    h->live_used = 0;
    h->live_count = 0;
    if (old != NULL) {
        for (int64_t i = 0; i < old_cap; i++) {
            if (old[i].state == 1) {
                int found = 0;
                int64_t idx = pcc_timer_live_probe(h, old[i].timer_id, &found);
                PccTimerLiveSlot *slot = &h->live[idx];
                slot->timer_id = old[i].timer_id;
                slot->deadline = old[i].deadline;
                slot->seq = old[i].seq;
                slot->state = 1;
                h->live_used++;
                h->live_count++;
            }
        }
        free(old);
    }
    return 0;
}

static int pcc_timer_live_reserve(PccTimerHeap *h) {
    /* Keep load factor under 0.7 counting tombstones. */
    if ((h->live_used + 1) * 10 <= h->live_cap * 7) return 0;
    int64_t new_cap = h->live_cap;
    /* Grow if live entries are dense; else a same-size rehash purges
     * tombstones. */
    if ((h->live_count + 1) * 10 > h->live_cap * 5) {
        new_cap = h->live_cap * 2;
    }
    return pcc_timer_live_rehash(h, new_cap);
}

/* ---- min-heap ----------------------------------------------------------- */

/* Order: deadline, then seq. Returns <0 if a<b, 0 if equal, >0 if a>b. */
static int pcc_timer_node_cmp(
    const PccTimerHeapNode *a,
    const PccTimerHeapNode *b
) {
    if (a->deadline < b->deadline) return -1;
    if (a->deadline > b->deadline) return 1;
    if (a->seq < b->seq) return -1;
    if (a->seq > b->seq) return 1;
    return 0;
}

static int pcc_timer_heap_reserve(PccTimerHeap *h) {
    if (h->heap_len < h->heap_cap) return 0;
    int64_t new_cap = h->heap_cap > 0 ? h->heap_cap * 2 : PCC_TIMER_HEAP_MIN_CAP;
    PccTimerHeapNode *grown = (PccTimerHeapNode *)realloc(
        h->nodes, (size_t)new_cap * sizeof(PccTimerHeapNode)
    );
    if (grown == NULL) return -1;
    h->nodes = grown;
    h->heap_cap = new_cap;
    return 0;
}

static void pcc_timer_sift_up(PccTimerHeap *h, int64_t i) {
    PccTimerHeapNode item = h->nodes[i];
    while (i > 0) {
        int64_t parent = (i - 1) / 2;
        if (pcc_timer_node_cmp(&item, &h->nodes[parent]) >= 0) break;
        h->nodes[i] = h->nodes[parent];
        i = parent;
    }
    h->nodes[i] = item;
}

static void pcc_timer_sift_down(PccTimerHeap *h, int64_t i) {
    int64_t n = h->heap_len;
    PccTimerHeapNode item = h->nodes[i];
    for (;;) {
        int64_t left = 2 * i + 1;
        int64_t right = left + 1;
        int64_t smallest = i;
        const PccTimerHeapNode *best = &item;
        if (left < n && pcc_timer_node_cmp(&h->nodes[left], best) < 0) {
            smallest = left;
            best = &h->nodes[left];
        }
        if (right < n && pcc_timer_node_cmp(&h->nodes[right], best) < 0) {
            smallest = right;
            best = &h->nodes[right];
        }
        if (smallest == i) break;
        h->nodes[i] = h->nodes[smallest];
        i = smallest;
    }
    h->nodes[i] = item;
}

static void pcc_timer_heap_pop_root(PccTimerHeap *h) {
    /* Remove nodes[0]; move last into root and sift down. */
    h->heap_len--;
    if (h->heap_len > 0) {
        h->nodes[0] = h->nodes[h->heap_len];
        pcc_timer_sift_down(h, 0);
    }
}

/* True if the current heap root is stale (no matching live entry). */
static int pcc_timer_root_is_stale(const PccTimerHeap *h) {
    const PccTimerHeapNode *root = &h->nodes[0];
    int found = 0;
    int64_t idx = pcc_timer_live_probe(h, root->timer_id, &found);
    if (!found) return 1;
    const PccTimerLiveSlot *slot = &h->live[idx];
    return (slot->deadline != root->deadline || slot->seq != root->seq)
        ? 1
        : 0;
}

/* ---- public API --------------------------------------------------------- */

int pcc_timer_heap_init(PccTimerHeap *h) {
    if (h == NULL) return -1;
    memset(h, 0, sizeof(*h));
    h->live = (PccTimerLiveSlot *)calloc(
        PCC_TIMER_LIVE_MIN_CAP, sizeof(PccTimerLiveSlot)
    );
    if (h->live == NULL) return -1;
    h->live_cap = PCC_TIMER_LIVE_MIN_CAP;
    return 0;
}

void pcc_timer_heap_dispose(PccTimerHeap *h) {
    if (h == NULL) return;
    free(h->nodes);
    free(h->live);
    h->nodes = NULL;
    h->live = NULL;
    h->heap_len = 0;
    h->heap_cap = 0;
    h->live_count = 0;
    h->live_used = 0;
    h->live_cap = 0;
}

int pcc_timer_heap_insert(PccTimerHeap *h, int64_t deadline, int64_t timer_id) {
    if (h == NULL) return -1;
    if (pcc_timer_live_reserve(h) != 0) return -1;
    if (pcc_timer_heap_reserve(h) != 0) return -1;

    h->seq++;
    int64_t seq = h->seq;

    int found = 0;
    int64_t idx = pcc_timer_live_probe(h, timer_id, &found);
    PccTimerLiveSlot *slot = &h->live[idx];
    if (found) {
        /* Reschedule: overwrite the authoritative (deadline, seq); the old
         * heap node is now stale and will be skipped. */
        slot->deadline = deadline;
        slot->seq = seq;
    } else {
        slot->timer_id = timer_id;
        slot->deadline = deadline;
        slot->seq = seq;
        slot->state = 1;
        h->live_used++;
        h->live_count++;
    }

    int64_t i = h->heap_len;
    h->nodes[i].deadline = deadline;
    h->nodes[i].seq = seq;
    h->nodes[i].timer_id = timer_id;
    h->heap_len++;
    pcc_timer_sift_up(h, i);
    return 0;
}

int pcc_timer_heap_cancel(PccTimerHeap *h, int64_t timer_id) {
    if (h == NULL || h->live_count == 0) return 0;
    int found = 0;
    int64_t idx = pcc_timer_live_probe(h, timer_id, &found);
    if (!found) return 0;
    /* Lazy: mark dead in the live map only; the heap slot is skipped later. */
    h->live[idx].state = 2;
    h->live_count--;
    /* live_used stays: tombstone still occupies a probe slot until rehash. */
    return 1;
}

int pcc_timer_heap_is_registered(const PccTimerHeap *h, int64_t timer_id) {
    if (h == NULL || h->live_count == 0) return 0;
    int found = 0;
    (void)pcc_timer_live_probe(h, timer_id, &found);
    return found ? 1 : 0;
}

int64_t pcc_timer_heap_size(const PccTimerHeap *h) {
    return h == NULL ? 0 : h->live_count;
}

int pcc_timer_heap_peek(PccTimerHeap *h, int64_t *out_deadline) {
    if (h == NULL) return 0;
    while (h->heap_len > 0) {
        if (pcc_timer_root_is_stale(h)) {
            pcc_timer_heap_pop_root(h);
            continue;
        }
        if (out_deadline != NULL) *out_deadline = h->nodes[0].deadline;
        return 1;
    }
    return 0;
}

int64_t pcc_timer_heap_pop_expired(
    PccTimerHeap *h,
    int64_t now,
    int64_t *out_ids,
    int64_t out_cap
) {
    if (h == NULL) return 0;
    int64_t drained = 0;
    while (h->heap_len > 0) {
        if (pcc_timer_root_is_stale(h)) {
            /* Stale: cancelled or superseded. Drop without returning. */
            pcc_timer_heap_pop_root(h);
            continue;
        }
        if (h->nodes[0].deadline > now) {
            /* Root is in the future -> nothing else is due. */
            break;
        }
        /* If the caller passed a buffer, stop when it is full so the due id
         * stays registered for the next call rather than being lost. */
        if (out_ids != NULL && drained >= out_cap) break;

        int64_t timer_id = h->nodes[0].timer_id;
        /* Unregister (root retention ends on expiry). */
        int found = 0;
        int64_t idx = pcc_timer_live_probe(h, timer_id, &found);
        if (found) {
            h->live[idx].state = 2;
            h->live_count--;
        }
        pcc_timer_heap_pop_root(h);
        if (out_ids != NULL) out_ids[drained] = timer_id;
        drained++;
    }
    return drained;
}

File: pcc/py_runtime/py/py_timer_heap.py
"""pcc-Python port of py_timer_heap.c (scalable vthread timer structure).

Mirror of the C runtime structure ``pcc/py_runtime/src/py_timer_heap.c`` and of
the CPU-only oracle ``pcc/vthread/timer_oracle.py::MinHeapTimerQueue``. All
three implement the SAME algorithm so they can be diffed against one another:

  * a binary min-heap keyed on ``(deadline, seq)`` (insert O(log n), pop O(log
    n), peek O(1)); and
  * lazy cancellation via an authoritative ``timer_id -> (deadline, seq)`` live
    map: ``cancel`` is O(1) amortized and only marks the id dead; the stale
    heap slot is skipped when it surfaces at the root during ``pop_expired``.

The dependency-free C twin is now embedded in the production scheduler in
``pcc_threads.c`` and has replaced its O(n)-insert sorted timer list (see
``docs/design/pcc-vthread-oracles.md``). This pcc-Python class remains the
executable semantic mirror: the production archive uses the C twin as its
low-level heap helper while all parked-thread ownership stays in the one
slot-based scheduler-root-handle contract.

Faithfulness note
-----------------
This port is written in the pcc-Python subset (classes + list + dict + int),
which is also valid CPython, so a host test can exercise it directly against
the oracle exactly like ``tests/vthread/test_timer_oracle.py`` exercises the
oracle. The C mirror keeps its own hand-rolled binary heap + open-addressing
live map (the pcc-Python heap here leans on ``list`` sift primitives written
out longhand so the pcc frontend does not need ``heapq``). Both preserve the
same observable semantics:

  * expiry order nondecreasing by deadline, FIFO among equal deadlines;
  * ``pop_expired(now)`` drains every id with ``deadline <= now`` (inclusive);
  * root retention until expired or cancelled;
  * cancelled / superseded entries are dropped without being returned.
"""

__pcc_runtime_port__ = True


class MinHeapTimerQueue:
    """Binary min-heap timer queue with lazy cancellation."""

    def __init__(self) -> None:
        # Parallel arrays form the heap (avoids tuple allocation churn and maps
        # cleanly onto the C node array). Index i holds one timer node.
        self._deadline = []  # type: list
        self._seq = []  # type: list
        self._timer_id = []  # type: list
        # Authoritative live registration: timer_id -> [deadline, seq].
        self._live = {}  # type: dict
        self._seq_counter = 0

    # -- introspection -----------------------------------------------------

    def size(self) -> int:
        return len(self._live)

    def is_registered(self, timer_id: int) -> bool:
        return timer_id in self._live

    # -- heap primitives ---------------------------------------------------

    def _less(self, i: int, j: int) -> bool:
        di = self._deadline[i]
        dj = self._deadline[j]
        if di < dj:
            return True
        if di > dj:
            return False
        return self._seq[i] < self._seq[j]

    def _swap(self, i: int, j: int) -> None:
        d = self._deadline[i]
        self._deadline[i] = self._deadline[j]
        self._deadline[j] = d
        s = self._seq[i]
        self._seq[i] = self._seq[j]
        self._seq[j] = s
        t = self._timer_id[i]
        self._timer_id[i] = self._timer_id[j]
        self._timer_id[j] = t

    def _sift_up(self, i: int) -> None:
        done = False
        while i > 0 and not done:
            parent = (i - 1) // 2
            if self._less(i, parent):
                self._swap(i, parent)
                i = parent
            else:
                done = True

    def _sift_down(self, i: int) -> None:
        n = len(self._deadline)
        done = False
        while not done:
            left = 2 * i + 1
            right = left + 1
            smallest = i
            if left < n and self._less(left, smallest):
                smallest = left
            if right < n and self._less(right, smallest):
                smallest = right
            if smallest == i:
                done = True
            else:
                self._swap(i, smallest)
                i = smallest

    def _pop_root(self) -> None:
        last = len(self._deadline) - 1
        self._swap(0, last)
        # Drop the (old-root) tail entry.
        del self._deadline[last]
        del self._seq[last]
        del self._timer_id[last]
        if len(self._deadline) > 0:
            self._sift_down(0)

    def _root_is_stale(self) -> bool:
        tid = self._timer_id[0]
        if tid not in self._live:
            return True
        entry = self._live[tid]
        if entry[0] != self._deadline[0]:
            return True
        if entry[1] != self._seq[0]:
            return True
        return False

    # -- mutation ----------------------------------------------------------

    def insert(self, deadline: int, timer_id: int) -> None:
        """Register timer_id to fire at deadline (O(log n)).

        Re-inserting a registered id reschedules it: the live entry is
        overwritten and the old heap node becomes stale (skipped later).
        """
        self._seq_counter = self._seq_counter + 1
        seq = self._seq_counter
        self._live[timer_id] = [deadline, seq]
        i = len(self._deadline)
        self._deadline.append(deadline)
        self._seq.append(seq)
        self._timer_id.append(timer_id)
        self._sift_up(i)

    def cancel(self, timer_id: int) -> bool:
        """Cancel a registered id (O(1) amortized). Returns True if it was live.

        Lazy: only the live entry is removed; the stale heap slot is skipped.
        """
        if timer_id in self._live:
            del self._live[timer_id]
            return True
        return False

    def peek(self, out) -> bool:
        """Peek the soonest live deadline into out[0]. Returns True if any.

        May pop leading stale slots first (amortized O(1)).
        """
        result = False
        done = False
        while len(self._deadline) > 0 and not done:
            if self._root_is_stale():
                self._pop_root()
            else:
                if len(out) > 0:
                    out[0] = self._deadline[0]
                result = True
                done = True
        return result

    def pop_expired(self, now: int) -> list:
        """Return ids with deadline <= now in expiry order.

        Nondecreasing by deadline, FIFO among equal deadlines. Stale
        (cancelled / superseded) slots are skipped. Expired ids are
        unregistered (root retention ends).
        """
        out = []
        done = False
        while len(self._deadline) > 0 and not done:
            if self._root_is_stale():
                self._pop_root()
            elif self._deadline[0] > now:
                done = True
            else:
                tid = self._timer_id[0]
                del self._live[tid]
                self._pop_root()
                out.append(tid)
        return out
